### DISCUSSION TITLE
Updating to 3.0 and getting "already loaded" error: `can't be configured using "with"`

### DISCUSSION THREAD

**Discussion by fwextensions** [uswds/uswds#4854] [Q&A] [answered]
I'm helping with an OSS project that uses USWDS 2.13.3 and I'm trying to update to 3.0.  I haven't used USWDS before, and Sass only lightly, so I seem to be out of my depth.

Following the docs, I created a new `theme/styles.scss` file that contains:

```sass
// Include a USWDS settings file (required)
@forward 'uswds-theme';

// Point to the USWDS source code (required)
@forward '../node_modules/@uswds/uswds';

// Include your project's custom Sass (optional)
@forward 'uswds-theme-custom-styles';
```

The project doesn't use Gulp to build the Sass, instead using the sass CLI directly:

```
sass --load-path=node_modules/@uswds/uswds/packages --no-source-map --style=compressed theme/styles.scss build/libraries/uswds/theme/styles.css
```

The `_uswds-theme.scss` file just contains an `@use` directive with custom settings migrated from the 2.0 files.  Compiling it triggers this output:

```
Error: This module was already loaded, so it can't be configured using "with".
   ┌──> node_modules/@uswds/uswds/packages/uswds-core/src/styles/variables/_index.scss
16 │   @forward "project-defaults";
   │   ^^^^^^^^^^^^^^^^^^^^^^^^^^^ new load
   ╵
   ┌──> node_modules/@uswds/uswds/packages/uswds-core/src/styles/functions/general/get-default.scss
14 │   @use "../../variables/project-defaults" as *;
   │   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ original load
   ╵
   ┌──> theme/_uswds-theme.scss
1  │ ┌ @use 'uswds-core' with (
2  │ │   $theme-image-path: './img',
3  │ │   $theme-show-compile-warnings: true,
4  │ │   $theme-show-notifications: false,
5  │ │   $theme-layout-grid-use-important: false,
6  │ │   $theme-global-border-box-sizing: true,
7  │ │   $theme-focus-color: 'blue-40v',
8  │ │   $theme-focus-offset: 0,
9  │ │   $theme-focus-style: solid,
10 │ │   $theme-focus-width: 0.5,
11 │ │   $theme-icon-image-size: 2,
12 │ │   $theme-breadcrumb-background-color: 'white',
13 │ │   $theme-site-max-width: 'desktop',
14 │ │   $theme-font-type-sans: 'public-sans',
15 │ │   $theme-font-weight-semibold: 600,
16 │ │ );
   │ └─^ configuration
   ╵
  node_modules/@uswds/uswds/packages/uswds-core/src/styles/variables/_index.scss 16:1  @forward 
  node_modules/@uswds/uswds/packages/uswds-core/src/styles/_index.scss 2:1             @forward 
  _index.scss 1:1                                                                      @use     
  theme/_uswds-theme.scss 1:1                                                          @forward 
  theme/styles.scss 2:1                                                                root stylesheet
```

This seems to be some issue internal to the USWDS source, as it sounds like it's using one internal file and then loading it again.  I'm not sure how to fix that.

<hr />

As often happens, asking this question helped me rubber-duck my way to some progress.  I was going to not include all the settings in the `with ()` statement, but then figured I should try commenting them out and then building.  That avoided this error and revealed another.  I eventually narrowed the issue down to this setting: `$theme-site-max-width: 'desktop'`

So... whut?  Why does that one setting break the build?  Not sure what it does, but it was in the existing settings.

Fixing that revealed a problem with the second `@forward` in the styles file, where the path wasn't right.  I guess I didn't read the docs closely enough, but this example is pretty misleading:

```
// Point to the USWDS source code (required)
@forward "./path/to/source/uswds";
```

It's not a path to the overall "source", which I assumed to be the NPM package.  Text lower down makes it clearer that it needs a path to a root stylesheet:

> When you install with npm, the complete path is typically ./node_modules/@uswds/uswds/packages/uswds/_index.html.

Though note that the documentation is pointing to a `.html` file, which is wrong.  I changed the line to `@forward '../node_modules/@uswds/uswds/packages/uswds';`, leaving off the index, and that seems to work. 

Now I'm dealing with an error that says `@include focus-shadow();` is undefined, when it clearly is, and used to work.  I'm guessing it's some Sass naming nonsense caused by using `@forward` instead of `@import`, but I still need to figure that out.

**Comment by amyleadem** [comment, marked as answer]
Hi @fwextensions,
Thanks for reaching out. It seems like you are well on your way to figuring this out, but I wanted to share an explanation about why you are seeing the configuration error. 

**Configuration error**
The `can't be configured using "with"` error can pop up when a non-USWDS variable is included inside `@use 'uswds-core' with()`. In this case, you pinpointed that the `$theme-site-max-width` variable is causing the error. This is because  this variable no longer exists in USWDS, having been replaced with`$theme-grid-container-max-width` in USWDS 2.11.2. So if you need to customize this setting, you can do so using `$theme-grid-container-max-width`. If you are happy with the default value, you can safely remove that line from your uswds-core configuration. 

**Focus-shadow() error**
USWDS has a `focus-outline()` mixin but I cannot find a`focus-shadow()` mixin. Do you see the `focus-shadow()` mixin elsewhere in your code base? If so, you will need to pull in the file via `@use [file_name] as *` at the top of your file. If not, let me know and I can circle back. 

**Next steps**
In case you wanted to follow along, here are the open issues related to this discussion:
- https://github.com/uswds/uswds-site/issues/1690
- https://github.com/uswds/uswds/issues/4864
- https://github.com/uswds/uswds-site/issues/1711

Thank you for sharing these issues with us, and I hope I answered your questions. Please feel free to follow-up if you still need assistance.

**Reply by fwextensions** [reply]
Thanks a lot for your help, and opening those doc fixes.  That should help folks in the future, especially that baffing Sass error when there are undefined variables.  

The main issue with the mixins is that this code base is redefining some that are from USWDS, like `h1`, `h2`, etc., which caused more Sass errors.  It's not just a simple matter of replacing `@import` with `@forward` everywhere.

So I went back to just importing the custom style file, while using `@forward` for USWDS in the `styles.scss` file.  That worked when running the sass build command locally, but it throws "file not found" errors when running in docker.  The commands and paths are set up the same, AFAICT.  The webpack sass loader also complains about not finding things when running in docker.  

I'm giving up on updating to 3.0 for now, as it seems like it'll require a lot of Sass debugging and rewriting for not much gain.  Maybe someone with more Sass-fu can figure it out at some point.

**Reply by mejiaj** [reply]
@fwextensions I don't think this problem with `@import` and `@forward` is exclusive to the design system. Those are new features that SASS modules uses to ensure there's no code duplication.

For example 

>Focus-shadow() error
USWDS has a focus-outline() mixin but I cannot find afocus-shadow() mixin. Do you see the focus-shadow() mixin elsewhere in your code base? If so, you will need to pull in the file via @use [file_name] as * at the top of your file. If not, let me know and I can circle back.

We've experienced similar errors. The way of dealing with this is importing the functions/mixins/variables to each partial that calls it. Here's a quick example with our `_body.scss` styles.

```scss
@use "uswds-core/src/styles/mixins/helpers/set-text-and-bg" as *;

body {
  @include set-text-and-bg($context: "Body");
  overflow-x: hidden;
}
```
https://github.com/uswds/uswds/blob/304147aea1d99a39eadf8da92e36140cb83acbc0/packages/uswds-elements/src/styles/_body.scss#L1-L7

You can see we're importing the `set-text-and-bg` mixin and then calling it below. You'll have to do something similar when you move your code base with custom styles to SASS modules. You might want to look into SASS' migrator tool to get a clearer picture - https://sass-lang.com/documentation/cli/migrator.

**Comment by ryparker** [comment]
For anyone reaching this migration thread now, a working USWDS 3 setup with the Sass CLI is:

```scss
// theme/styles.scss
@forward "uswds-theme";
@forward "uswds";
@forward "uswds-theme-custom-styles";
```

```sh
sass --load-path=node_modules/@uswds/uswds/packages theme/styles.scss build/libraries/uswds/theme/styles.css
```

Configure `uswds-core` in `_uswds-theme.scss` before loading the rest of USWDS. The current setting is `$theme-grid-container-max-width`.

Each custom partial also needs its own imports. Forwarding USWDS from `styles.scss` does not put its mixins in every partial's scope. A namespace helps avoid collisions with project mixins:

```scss
// theme/_uswds-theme-custom-styles.scss
@use "uswds-core" as uswds;

.example:focus {
  @include uswds.focus-outline;
}
```

I compiled this structure against USWDS 3.14.0. It addresses the entry-point and mixin-scope questions. The Docker-only “file not found” error needs the specific missing import to diagnose: check the command's working directory, files copied into the image, and filename case inside the container. The information in this thread does not establish which of those caused that separate failure.
